src/StringTools/StringTools.cs

```diff
@@ -0,0 +1,232 @@
+        /// </summary>
+        private static object s_locker = new object();
+
+
+        #region Public methods
+
+        /// <summary>
+        /// Interns the given string if possible.
+        /// </summary>
+        /// <param name="str">The string to intern.</param>
+        /// <returns>The interned string equal to <paramref name="str"/>.</returns>
+        /// <remarks>
+        /// <see cref="TryInternStringDelegate"/> callbacks registered via <see cref="RegisterStringInterningCallback"/> are called
+        /// in registration order to check if custom interning is desired. If no callback returns true, the default interner based
+        /// on weak GC handles is used.
+        /// </remarks>
+        public static string TryIntern(string str)
+        {
+            InternableString internableString = new InternableString(str);
+            return OpportunisticIntern.Instance.InternableToString(ref internableString);
+        }
+
+#if !NET35
+        /// <summary>
+        /// Interns the given readonly character span if possible.
+        /// </summary>
+        /// <param name="str">The character span to intern.</param>
+        /// <returns>The interned string equal to <paramref name="str"/>.</returns>
+        /// <remarks>
+        /// <see cref="TryInternStringDelegate"/> callbacks registered via <see cref="RegisterStringInterningCallback"/> are called
+        /// in registration order to check if custom interning is desired. If no callback returns true, the default interner based
+        /// on weak GC handles is used.
+        /// </remarks>
+        public static string TryIntern(ReadOnlySpan<char> str)
+        {
+            InternableString internableString = new InternableString(str);
+            return OpportunisticIntern.Instance.InternableToString(ref internableString);
+        }
+#endif
+
+        /// <summary>
+        /// Returns a new or recycled <see cref="RopeBuilder"/>.
+        /// </summary>
+        /// <returns>The RopeBuilder.</returns>
+        /// <remarks>
+        /// Call <see cref="IDisposable.Dispose"/> on the returned instance to recycle it.
+        /// </remarks>
+        public static RopeBuilder GetRopeBuilder()
+        {
+#if NET35
+            return new RopeBuilder();
+#else
+            return s_ropeBuilderPool.Get();
+#endif
+        }
+
+        /// <summary>
+        /// Adds a callback to be called when a string is being interned.
+        /// </summary>
+        /// <param name="callback">The callback to add.</param>
+        /// <remarks>
+        /// Use this to implement custom interning for some strings. The callback has access to the string being interned
+        /// via the <see cref="InternableString"/> representation and can override the default behavior by returning a custom
+        /// string.
+        /// </remarks>
+        public static void RegisterStringInterningCallback(TryInternStringDelegate callback)
```

Comment:
This callback scheme is a complexity it would be good to avoid. The decision to hard code a lookup for short well known strings was based on measurements a long time ago, and also in a context where there was no callback involved. I suggest to rip it out and do your measurements with no strings propopulated. You can look at the stats dump you have built in to see which strings are very common (they may have changed since this list was chosen). Then you can decide whether the data suggests they need special handling. If they do that you might not choose the same callback-and-switch scheme but something else.

Reply:
That's a great point. The callback used by MSBuild has both "positive" and "negative" strings and I was more worried about the latter, i.e. strings that must not be interned. It doesn't look like we actually depend on this in any way, though, and it's probably just a left-over from a logic removed long ago.

I'll take detailed measurements and will get back on the value of hardcoded strings. Thank you!

Reply:
I have removed the callback, you were right that hardcoded strings were not helping. Thank you!